airflow-core/src/airflow/api_fastapi/core_api/routes/public/task_instances.py

```diff
@@ -181,7 +181,11 @@ def get_mapped_task_instances(
     # 0 can mean a mapped TI that expanded to an empty list, so it is not an automatic 404
     unfiltered_total_count = get_query_count(query, session=session)
     if unfiltered_total_count == 0:
-        dag = dag_bag.get_dag(dag_id)
+        dag_run = session.scalar(select(DagRun).where(DagRun.dag_id == dag_id, DagRun.run_id == dag_run_id))
+        if dag_run:
+            dag = dag_bag.get_dag_for_run(dag_run, session=session)
+        else:
+            dag = dag_bag.get_latest_version_of_dag(dag_id, session=session)
```

Comment:
We use this pattern in couple of places, maybe extract it out as a helper function that takes a dag_id, run_id and session?

Reply:
Should we do it as part of https://github.com/apache/airflow/pull/53153#pullrequestreview-3050567892?